### ISSUE TITLE
Case study: Create velocity field maps

### ISSUE THREAD

**Issue by dabokun** [aflak-vis/aflak#22]
銀河の画像の各位置(pixel)におけるスペクトルから、それぞれの位置におけるドップラー効果のズレを計算し、視線方向に対する銀河の運動速度のマップを作成する。

機構としては
①3次元画像とある波長範囲を指定する。(入力)
②Gaussian-Fitting等を利用し、結果からズレの度合いを計算。
③速度マップを生成。

**Comment by makotouemura** [comment]
検討して頂き、ありがとうございます。実現すれば直ちに天文学的な成果がでる、というものではありませんので、優先順位はお任せします。

いきなり Gaussian-Fitting は難しい/計算時間がかかる、かもしれないので、その場合は、単に「指定された波長範囲内での最大値の波長」とか「重心」（＝ (sum_i f_i * l_i)/(sum_i f_i)；f_i, l_i はi番目のフラックスと波長）とかでも良いと思います。要するに輝線の代表的な波長が得られれば良いので。  

波長から速度に変換する際には基準となる波長が必要です。これは輝線の静止波長とするか、銀河の平均速度とするか、ですが、この値も入力にしてしまうのが良いかもしれません。

**Comment by makotouemura** [comment]
はい、それでOKです。画像全体のスペクトルの代表輝線の波長を l、その輝線の静止波長を l0 とした時、天体の視線速度は v = c*(l-l0)/l0 (cは光速 = 3e5 km/s）です。ただしこれは(l-l0)/l0 の値が小さい場合にのみ成り立つ近似式です。
参考：[http://www.th.phys.titech.ac.jp/~muto/lectures/INP02/INP02_chap11.pdf](http://www.th.phys.titech.ac.jp/~muto/lectures/INP02/INP02_chap11.pdf)
の式11.13, 11.14。

**Comment by dabokun** [comment]
相対論の効果ですね。
分かりました。それでやってみます。何か動くものができましたら、またここに投稿します。

**Comment by kazuyakusastro** [comment]
横槍で失礼します。
銀河内の速度差はほとんどの場合±1000 km/sよりも小さいため、l0を銀河の平均速度での波長として、v = c*(l-l0)/l0で計算して大丈夫です。

**Comment by dabokun** [comment]
こんにちは。

とりあえず、
・ある波長範囲でfluxが最大値/最小値となる波長で構成されるマップ
・ある波長範囲での重心波長で構成されるマップ
と、シフトしたあとの画像全体の代表波長(今回はHαです)を使って、画像全体の動きに対する各位置の相対速度のマップを作成しました。

実際の天文学者の方々の作成例と比較しなければ評価はできませんが、作ってみて考えたことを以下に記載します：
・最大値/最小値で作成した場合、離散的な波長の値を返すので、やはり前後の値などを使ってフィッティングすることが良いと考えます。
・現在はfluxの最大値/最小値をとる波長のマップのみを返していますが、fluxの最大値/最小値そのもののマップを返す必要はありますか？これは必要であれば簡単に実装できます。
・ノードの名前などがわかりにくいところがあるかもしれません。

他にも、何かありましたらコメントをお願いします。
![image](https://user-images.githubusercontent.com/37533832/49640673-82e2d980-fa51-11e8-9a73-9e7c21253b4c.png)

**Comment by makotouemura** [comment]
あと「fluxの最大値/最小値そのもののマップ」は表示できると良いです。ただし、輝線の強度は「等価幅マップ」で見る事の方が多いです。

**Comment by malikolivier** [comment]
> 今はこのような特定の空間領域だけのスペクトルを示すことはできるのでしょうか？

1次元データ（e.g. スペクトル）のoutput windowの中でmouseをwheelを使えばzoom in/zoom outができます。それで十分でしょうか？ また、現在1次元データをclipするノードが存在しません。必要であればそういったノードを作っても問題ないと思います。

それ以外の質問について、 @dabokun にバトンを渡します。

**Comment by makotouemura** [comment]
あ、わかりました。extract_wave の ROI を Whole image じゃなくて、特定の空間領域を指定する事で、その部分だけのスペクトルが表示されるんですね？　それなら zoom in/zoom out があれば十分だと思います。

**Comment by malikolivier** [comment]
@makotouemura ROIを作るには、2次元のoutput windowの画像に右クリックをして、ROIを選択すれば、ROIノードが作成されます。pixelを一つ一つ選択することによって、ROIを構築できます。 ROIを拡張する予定です（ #29）。

![image](https://user-images.githubusercontent.com/14120117/49682037-d3a90f80-faef-11e8-8440-41d4aff3372d.png)

ちなみに、今aflakの最新版はmacOSをちゃんと対応しているはずなので、既にそうでなかれば、updateしてください。update方法はREADMEに記載されています。

**Comment by malikolivier** [comment]
@dabokun master に #44 をマージしました。このユースケースを代表するtemplateは作れる？

**Comment by dabokun** [comment]
@malikolivier https://github.com/aflak-vis/aflak/pull/49 でtemplateを作成してみました。
